### ISSUE TITLE
What exactly is the difference between a parametric and non-parametric model?

### ISSUE THREAD

**Issue by mycroft1018** [mycroft1018/Econ_101#4]
The naming is really bad: parametric models should better be called "models with fixed number of parameters", non-parametric models should better be called "models with unbound number of parameters".

**Comment by mycroft1018** [comment]
In a parametric model, the number of parameters is fixed with respect to the sample size. In a nonparametric model, the (effective) number of parameters can grow with the sample size.

In an OLS regression, the number of parameters will always be the length of β, plus one for the variance.

A neural net with fixed architecture and no weight decay would be a parametric model.

But if you have weight decay, then the value of the decay parameter selected by cross-validation will generally get smaller with more data. This can be interpreted as an increase in the effective number of parameters with increasing sample size.

**Comment by mycroft1018** [comment]
I think that the word "effective" in the accepted answer should be deleted. Because due to the different number of effective parameters, as [Aksakal pointed out](https://stats.stackexchange.com/questions/268638/what-exactly-is-the-difference-between-a-parametric-and-non-parametric-model/268646#comment709412_268646), the accepted answer implies that Ridge and Lasso are non-parametric, but it doesn't seem to be true. Effective parameters (effective degrees of freedom) are characteristics of a learning algorithm, but not a model itself.

In a machine learning problem we have three things:

Data generation model. It describes our assumptions about the probabilistic distribution that generated our data. From mathematical statistics we know that data generation model can be parametric or non-parametric. As [Glen_b pointed out](https://stats.stackexchange.com/questions/268638/what-exactly-is-the-difference-between-a-parametric-and-non-parametric-model/268646#comment513816_268638), in parametric data generation model this distribution is defined by a fixed number of parameters. In nonparametric data generation model we don't have a distribution with a fixed number of parameters, we have milder assumptions about it, like continuity or symmetry.

Algorithm (hypothesis). It is a function h:X→Y
 from some hypothesis space H
. This function tries to predict the true target value on any sample x
. Hypothesis space (model) can be parametric or non-parametric.
In parametric hypothesis space (parametric model) every algorithm is uniquely defined by a fixed number of parameters (this number is the same for all algorithms from this space). Simple examples of parametric models are linear regression model: H={h(x;w,b)=⟨w,x⟩+b∣w∈Rd,b∈R}

and linear (binary) classification model: H={h(x;w,b)=sign(⟨w,x⟩+b)∣w∈Rd,b∈R}
.
In non-parametric models we can't describe every algorithm with the vector of parameters of the same constant size for all algorithms. Usually the number of parameters grows with the size of a training set. For example in the case of decision trees H={T(x;Θ)∣Θ}
, where Θ={J,{Rj,γj}Jj=1}
 is a vector of tree's parameters: J
 is a number of terminal nodes in the tree, Rj
 are subregions of the input space X
 corresponding to these nodes, and γj
 are the predictions in these nodes. Trees can have different number of leaves and different number of internal nodes, so the space of decision trees is non-parametric (dimension of Θ
 will be different for different trees, if we train them on the datasets generated from the same distribution, that is, with the same number of features d
, but with different number of observations in each dataset).

Method (learning algorithm). We can formalize it as a function μ:D→H
. It uses training set D
 to fit some hypothesis h∈H
. If H
 is parametric we call μ
 parametric method. If H
 is non-parametric we call μ
 non-parametric method. For example, OLS, Ridge and Lasso are all parametric methods because they all use exactly the same parametric model H={h(x;w,b)=⟨w,x⟩+b∣w∈Rd,b∈R}
 (as I said above, we call it "linear regression model"). Despite the fact that these methods have different number of effective parameters.

Keep in mind, that we can use parametric data generation model and non-parametric learning algorithm (or vice-versa). For example, we can have gaussian data generation model Y=f(X)+ε
, where ε∈N(0,σ2)
. Obviously, this is a parametric data generation model. But we can always fit non-parametric method (for example, kNN regression) on the training set D
, generated by this model.
Similarly, we can fit parametric OLS method without any parametric assumptions about data generation process. In this case this method simply will not be equivalent to Maximum Likelihood Estimation.

**Comment by mycroft1018** [comment]
[https://stats.stackexchange.com/questions/268638/what-exactly-is-the-difference-between-a-parametric-and-non-parametric-model](url)
